### ISSUE TITLE
Run distributed key generation on-chain, storing `Ack`s and `Part`s in a contract.

### ISSUE THREAD

**Issue by afck** [poanetwork/parity-ethereum#35]
Whenever the set of validators changes (due to a new staking epoch or a faulty node being removed), the new validators need to generate their threshold keys, so Honey Badger can be restarted with the new configuration.

To do this, each of the new validators needs to create a `SyncKeyGen` instance with all new validators' public key, and publish their `Part` message on the contract. After each block, they handle all key gen messages stored on the contract. After the first block with which `SyncKeyGen::count_complete` is at least _N - f_—i.e. ⅔ of the new validators' parts are complete—, the new keys are generated. All old validators terminate their Honey Badger instances, and the new ones start theirs, beginning at the next block.

**Comment by afck** [comment]
@varasev: For this we would need an additional smart contract—not _too_ smart, though: It just needs to record the history of messages that were exchanged during the key generation process.

Validators—and only the ones that were selected for the next staking epoch—can commit one `Part` message and up to _N_ (number of validators) `Ack` messages each. They should probably be stored as events containing the serialized messages. The contract could even be oblivious of the message types. If that's useful, we could make it a bit smarter, though: It could count the number of `Ack`s referring to each `Part`, and know that key generation is complete when more than ⅔ of the `Part`s received `Acks` from at least ⅓ of the validators. After a block where key generation completes, a new Honey Badger instance with the new validator set will be started.

**Comment by afck** [comment]
Actually: Can Parity record events in blocks even without a specific contract? Then we could just store the key generation messages directly.

**Comment by varasev** [comment]
Please take a look at the draft of the contract: https://github.com/poanetwork/pos-contracts/blob/master/contracts/KeyGenHistory.sol

Is this approximately what we need?

**Comment by afck** [comment]
Looks great!
I think `/ 3` is wrong, though: Let's multiply by `3` on the other side, so it doesn't round down.

**Comment by afck** [comment]
Also, we'll probably need events for the messages: It should be possible for a restarting node to read the key generation messages from the blockchain and reconstruct its key shares.

**Comment by varasev** [comment]
Ok, I've made corresponding changes: https://github.com/poanetwork/pos-contracts/blob/master/contracts/KeyGenHistory.sol

The `PartWritten` and `AckWritten` events were added. Are they correct?

**Comment by varasev** [comment]
Please pay attention to the next things:

- `writePart` function should be called before `writeAck` is called: https://github.com/poanetwork/pos-contracts/blob/6a7121ae16f34f6d0b40e4b1b774dd3bb14bbbfa/contracts/KeyGenHistory.sol#L53

- `writePart` function can only be called once by the specified validator in the current staking epoch: https://github.com/poanetwork/pos-contracts/blob/6a7121ae16f34f6d0b40e4b1b774dd3bb14bbbfa/contracts/KeyGenHistory.sol#L41

- `writeAck` function can only be called once for the same `ack` message in the current staking epoch by the specified validator: https://github.com/poanetwork/pos-contracts/blob/6a7121ae16f34f6d0b40e4b1b774dd3bb14bbbfa/contracts/KeyGenHistory.sol#L54 - i.e., the validator cannot call `writeAck` twice for the same `ack` message in the current staking epoch.

**Comment by afck** [comment]
Thanks, that looks great, and these three requirements make perfect sense. :+1: 

I assume `changeRequestCount` is a counter for validator changes (banning faulty nodes) during a staking epoch, and we'd reset it to zero in each new staking epoch?

**Comment by varasev** [comment]
> I assume `changeRequestCount` is a counter for validator changes (banning faulty nodes) during a staking epoch, and we'd reset it to zero in each new staking epoch?

I propose to simply increment `changeRequestCount` each time when a validator set changes (despite `stakingEpoch` counter) and not to reset it. But if that's necessary, we could reset as you suggested.

**Comment by afck** [comment]
You're right, always incrementing is simpler. :+1:

**Comment by afck** [comment]
I think there's one more thing we'll have to store on the blockchain (probably in `ReportingValidatorSet`?), namely the validators'/candidates' (serialized) [`threshold_crypto::PublicKey`](https://docs.rs/threshold_crypto/0.2.1/threshold_crypto/struct.PublicKey.html)s. These are needed to initialize key generation, and also to start Honey Badger.

**Comment by afck** [comment]
(On the other hand, it would also be feasible to exchange these via Whisper, so they don't necessarily need to go into the contract.)

**Comment by varasev** [comment]
We could add the next function for that:

```solidity
mapping(address => bytes) public publicKey;

function savePublicKey(address _validator, bytes _key) public onlySystem {
    publicKey[_validator] = _key;
}
```

or (for batch saving)

```solidity
function savePublicKeys(address[] _validators, bytes[] _keys) public onlySystem {
    require(_validators.length == _keys.length);
    for (uint256 i = 0; i < _validators.length; i++) {
        publicKey[_validators[i]] = _keys[i];
    }
}
```

Is this correct?

Should there be any function to remove the key from the contract after its using? E.g.:

```solidity
function removePublicKey(address _validator) public onlySystem {
    delete publicKey[_validator];
}
```

**Comment by afck** [comment]
Yes, that would be perfect!
I don't think we'd batch-add public keys. The validators would rather publish theirs individually, probably as soon as they register? I also don't think we'll ever need to remove them. Possibly update, but for that, `savePublicKey` would do.

**Comment by varasev** [comment]
Ok, the `savePublicKey` function will be called by SYSTEM_ADDRESS, right?

**Comment by afck** [comment]
No, I think it would be called by the validator themselves.
Ideally they should include their public key as soon as they register with the contract as validators in the first place.

**Comment by varasev** [comment]
Ok, `savePublicKey` function will be called by validator's (observer's) Parity node and `msg.sender` will be the address of validator. If so, then `savePublicKey` function should look like the following:

```solidity
function savePublicKey(bytes _key) public {
    require(doesPoolExist(msg.sender));
    publicKey[msg.sender] = _key;
}
```

I.e., the `savePublicKey` should only be called by the existing observer.

The second possible variant is

```solidity
function savePublicKey(bytes _key) public {
    require(isValidator(msg.sender));
    publicKey[msg.sender] = _key;
}
```

I.e., the `savePublicKey` will succeed only if `msg.sender` is a validator from a new validator set, what is `true` after `newValidatorSet()` function is called or after `InitiateChange` event is emitted (both cases assume building the new set of validators).

**Comment by afck** [comment]
Sorry, I'm still confused about the notions "validator", "observer", "staker" and "pool". (Please also take a look at https://github.com/poanetwork/hbbft/issues/347 — let's make sure we find a consistent naming scheme for all layers of the stack.)
I think as soon as a node joins the network and registers itself with the contract as a _potential_ validator (i.e. it joins the pool?), it should also store its public key in the contract. There's no reason to delay its own (local, non-threshold) key generation until it becomes selected into a validator set, so ideally it wouldn't be possible to join the pool without also specifying one's public key.

(To clarify, this is a full public key, to which the node has the full secret key. These are different from the key _shares_ for _threshold_ encryption that are created in the distributed key generation process.)

**Comment by varasev** [comment]
Thanks for the link! In the terminology of the current [ReportingValidatorSet](https://github.com/poanetwork/pos-contracts/blob/676100bf0d0769089f8822e1a678c83c37e31951/contracts/ReportingValidatorSet.sol) contract:

- `observer` is the address which received at least one stake (observer can be validator if it is chosen randomly at the beginning of new staking epoch);
- `pool` is the same as `observer` (its another name);
- `staker` is the address which makes a stake for some observer (or a few observers). Staker put his stake in the pool of observer (pool of stakes for this observer).

Should I rename `observer` to something else?

**Comment by afck** [comment]
Thanks for the clarification! I wouldn't rename it just yet, but I'll quote your comment on the naming discussion thread.

With the notions as they are right now, I think that an observer should always publish their public key from the beginning. There shouldn't ever be an observer in the contract without a (BLS/threshold_crypto) public key.

**Comment by afck** [comment]
I.e. I imagine the public key should be a field in `ObserverState` (but I only just started reading the contract in detail, so I might misunderstand the meaning).

**Comment by varasev** [comment]
> I imagine the public key should be a field in ObserverState

Good idea, I'll think about that.

I've added the checking of public key existence when an address tries to become an observer: https://github.com/poanetwork/pos-contracts/commit/935cb03dee1158b7d491a1f91cec0172fd67e5ac

Will the public key have an exact length in bytes?

Can an address change its public key while it is an observer? E.g., it will call `savePublicKey` again: https://github.com/poanetwork/pos-contracts/blob/935cb03dee1158b7d491a1f91cec0172fd67e5ac/contracts/ReportingValidatorSet.sol#L112

**Comment by afck** [comment]
> Will the public key have an exact length in bytes?

Yes… I was going to say "48 bytes", but instead I opened [this issue](https://github.com/poanetwork/threshold_crypto/issues/63). :weary: 

> Can an address change its public key while it is an observer?

We don't have concrete plans for that yet, but I guess in general it's a good idea to allow them to periodically generate a new key?

**Comment by varasev** [comment]
> I guess in general it's a good idea to allow them to periodically generate a new key?

Yes, I think so too.

**Comment by varasev** [comment]
> We'll definitely fix that, of course, so it's going to be 48 bytes

Got it.

**Comment by varasev** [comment]
> I imagine the public key should be a field in `ObserverState`

The `publicKey` moved to the `ObserverState` struct: https://github.com/poanetwork/pos-contracts/commit/e125a7fe2f047145ad951d2a889c6e3d194a4edb

**Comment by c0gent** [comment]
@DemiMarie Please see my comments in Slack.

**Comment by afck** [comment]
@varasev: I think this line isn't right: https://github.com/poanetwork/pos-contracts/blob/98ec9196ceb911891f78bd200c8cb953f0d0c146/contracts/KeyGenHistory.sol#L68
And now that I think of it, I'm not sure if the whole counting and detecting completeness is even feasible within the contract. (It's not strictly necessary either; we can do that in the Rust code.)

Each `Ack` acknowledges a `Part`, usually from a different sender. So: Each of us calls `writePart`, and when I see your `PartWritten` events, I call `writeAck` with an `Ack` for each of them (including my own).
Unfortunately it's hard to tell which `Part` any given `Ack` belongs to: Its [first field](https://github.com/poanetwork/hbbft/blob/ceb416a6e145a46898637377e05ded4422a1494e/src/sync_key_gen.rs#L241) is the index of the `Part`'s sender, in a list where all the participanting nodes are sorted by their node ID.

It's probably not worth adding a lot of code for this: For our purposes it's enough if the contract just records the events, without any knowledge about the messages' meanings.
So `isKeyGenComplete` and most of the checks can probably be removed. The only thing the contract should enforce is that each of the _N_ participants writes at most 1 `Part` and _N_ `Ack`s.

Sorry for the confusion! I should have given more details in the issue in the first place.

**Comment by varasev** [comment]
No problem, I'll change the code of it.

**Comment by varasev** [comment]
@afck Please take a look: https://github.com/poanetwork/pos-contracts/blob/7febced6e1cdf292b83cca9b0388429544e95d7b/contracts/KeyGenHistory.sol - is it ok now?

**Comment by afck** [comment]
That looks good, thanks! :+1:

**Comment by varasev** [comment]
> There's an `InitiateChange` event now: [#33 (comment)](https://github.com/poanetwork/parity-ethereum/issues/33#issuecomment-444451125)
> A node that finds itself in `newSet` must start key generation and send its `Part`.

We agreed that we'll use the `getValidators()` getter instead of catching the event. Parity could just read the current validator set with `getValidators()` every block, and if it discovers that the set is changed in the contract, it launches changing of the set in the engine.

**Comment by afck** [comment]
Closed in favor of #157.
